### ISSUE TITLE
training of "Representation Learning of Questions" shows "Total Loss: nan"

### ISSUE THREAD

**Issue by jjkim** [oezyurty/KCQRL#7]
I ran the training of "Representation Learning of Questions" with the specified files of XES3G5M but it shows "Total Loss: nan".

**Comment by NousEU** [comment]
I'm having this issue as well, and the micro F1 score is the same for each epoch.

**Comment by RingoTC** [comment]
It casued by the computation of Contrastive Learning Loss. https://github.com/oezyurty/KCQRL/blob/main/representation_learning/trainer.py#L109

A fix from me could be

`def contrastive_loss(self, score_matrix, direct_pair_indices, clustered_pair_indices):
        pos_mask = torch.zeros_like(score_matrix)
        neg_mask = torch.ones_like(score_matrix)
        
        # We treat all the direct kc - quest/step pairs as positives
        pos_mask[direct_pair_indices[:, 1], direct_pair_indices[:, 0]] = 1
        neg_mask[direct_pair_indices[:, 1], direct_pair_indices[:, 0]] = 0
        
        # We discard all clustered kc - quest/step pairs as from negatives (in default mode)
        if not self.disable_clusters:
            neg_mask[clustered_pair_indices[:, 1], clustered_pair_indices[:, 0]] = 0
        
        pos_score = score_matrix * pos_mask
        neg_score = score_matrix * neg_mask

        pos_score_sum = pos_score.sum(dim=-1)
        neg_score_sum = neg_score.sum(dim=-1)

        # Check for any zeros or very small numbers in the denominator
        scores_sum = pos_score_sum + neg_score_sum

        # Safe-guarding against division by zero or log of zero
        scores_sum = scores_sum.clamp(min=1e-8)  # Avoid division by zero
        pos_score_sum = pos_score_sum.clamp(min=1e-8)  # Avoid log(0)

        # Calculate the loss for each element in pos_score
        ratio = pos_score / scores_sum.unsqueeze(-1) 
        ratio = torch.clamp(ratio, min=1e-8)  
        element_loss = -1 * torch.log(ratio)  

        # Mask out the elements that are zero in pos_score
        element_loss = element_loss * pos_mask

        # Calculate the mean loss for each row, considering only non-zero elements
        row_loss = element_loss.sum(dim=-1) / pos_mask.sum(dim=-1).clamp(min=1e-8)

        # Filter out rows where pos_mask has no positives
        valid_rows = pos_mask.sum(dim=-1) > 0 
        row_loss = row_loss[valid_rows]  # Filter to include only valid rows

        # Compute the mean loss over valid rows only
        cl_loss = row_loss.mean()

        return cl_loss
`
